README.md
# stacked_chart
With Shadow

<img width="494" alt="Screen Shot 2021-08-10 at 2 27 33 pm" src="https://user-images.githubusercontent.com/30488716/128808291-df9a5c1d-ee19-4360-96e2-51dd213698c6.png">



Without Shadow 

![animation](https://user-images.githubusercontent.com/30488716/127808681-453a915f-1b90-42b9-89e4-24c681fdb534.gif)


Animation for the onPressed callback

![onPress animation](https://user-images.githubusercontent.com/30488716/128816523-7c70c4a2-dbef-4527-bfc4-754807294c0d.gif)


A flutter package for creating a stack barchart with easy customization.


## Installation
In the `dependencies:` section of your `pubspec.yaml`, add the following line:

```yaml
stacked_chart:
```

Then `import` it as :

```yaml
import 'package:stacked_chart/stacked_chart.dart';
```

## Usage
```dart
import 'dart:math';

import 'package:flutter/material.dart';
import 'package:stacked_chart/stacked_chart.dart';

enum WeekDay { monday, tuesday, wednesday, thursday, friday, saturday, sunday }

class WeeklyChartDemo extends StatefulWidget {
  const WeeklyChartDemo({Key? key}) : super(key: key);

  @override
  _WeeklyChartDemoState createState() => _WeeklyChartDemoState();
}

class _WeeklyChartDemoState extends State<WeeklyChartDemo> {
  final rng = Random();
  List<BookingStatus> weeklyStatus = [];

  @override
  void initState() {
    createRandomWeeklyStatus();
    super.initState();
  }

  void createRandomWeeklyStatus() {
    weeklyStatus.clear();
    final List<BookingStatus> weeklyData = [];
    WeekDay.values
        .map(
          (day) => weeklyData.add(BookingStatus(
            dateTime: DateTime.now().add(
              Duration(days: day.index),
            ),
            bookings: {
              'Unfilled booking': rng.nextInt(20),
              'Filled booking': rng.nextInt(20),
            },
          )),
        )
        .toList();
    setState(() {
      weeklyStatus = weeklyData;
    });
  }

  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      onTap: createRandomWeeklyStatus,
      child: StackedChart(
        data: weeklyStatus,
        size: const Size(300, 150),
        showLabel: true,
      ),
    );
  }
}

class BookingStatus extends ChartData<LabelData, int>
    implements Comparable<BookingStatus> {
  final DateTime dateTime;
  final Map<String, int> bookings;

  static Map<LabelData, int> convertBookingToMapOfLabelDataInt(
      Map<String, int> bookings) {
    final Map<LabelData, int> convertedData = {};
    bookings.entries
        .map((e) => convertedData.addAll({LabelData(e.key): e.value}))
        .toList();
    return convertedData;
  }

  int get totalBookingCount =>
      bookings.values.reduce((total, value) => total = total + value);

  BookingStatus({required this.dateTime, this.bookings = const {}})
      : super(
            labelWithValue: convertBookingToMapOfLabelDataInt(bookings),
            barLabel: dateTime.day.toString());

  @override
  int compareTo(BookingStatus other) => dateTime.compareTo(other.dateTime);
}
```


CHANGELOG.md
## 0.0.6
* Added MainAxisAlignment to align the stack bar accordingly.
## 0.0.5
* Added OnPressed animation so that the bar animate as if you are pressing button
## 0.0.4
* fixed overflow issue when label is enabled
## 0.0.3
* Added optional `enableShadow` to toggle material and glass look for the bar
* Code clean up

## 0.0.2

* Added optional `labelStyle` to style the label.
* Moved example into main file.

## 0.0.1

* Initial project.


lib/src/data/data.dart
import 'dart:ui';

/// Contain Map of type K and V value and optional onPressed function
/// that can be used to call when the individual bar of barchart is
/// pressed
class ChartData<K extends LabelData, V extends num> {
  final Map<K, V> labelWithValue;
  final VoidCallback? onPressed;
  final Color? barBackGroundColor;
  final String barLabel;
  const ChartData({
    required this.labelWithValue,
    required this.barLabel,
    this.barBackGroundColor,
    this.onPressed,
  });

  /// Calculate the maximum value that is present in the value of type V
  V get maxSize => labelWithValue.isNotEmpty
      ? labelWithValue.values.reduce((total, value) => _combine(total, value))
      : 0 as V;

  /// Return sum of value with subsuquent value key. This is important to
  /// draw a bar in a stack as the subsequent value is smaller than the
  /// previous one creating stack of bars
  V getHeightValueForGivenKey(K key) {
    final labelWithValueWithOutPreviousValue = Map<K, V>.from(labelWithValue);
    for (final mapEntry in labelWithValue.entries) {
      if (mapEntry.key == key) {
        break;
      }
      labelWithValueWithOutPreviousValue.remove(mapEntry.key);
    }
    return labelWithValueWithOutPreviousValue.entries
        .map((e) => e.value)
        .reduce((value, element) => _combine(value, element));
  }

  /// Return the index for the given key for the[labelWithValue]
  int getIndexForGivenKey(K key) {
    int index = 0;
    labelWithValue.entries
        .toList()
        .asMap()
        .entries
        .map((e) => (e.value.key == key) ? (index = e.key) : 0)
        .toList();
    return index;
  }

  V _combine(V value, V element) => value + element as V;
}

/// Contain label and color for the chart
class LabelData {
  final String label;
  final Color? color;
  LabelData(this.label, [this.color]);
}

extension MaxSizeX on List<ChartData> {
  num get getMaxValue {
    num maxValue = 0;
    forEach((e) => e.maxSize > maxValue ? maxValue = e.maxSize : 0);
    return maxValue;
  }
}


lib/src/widgets/chart.dart
import 'package:flutter/material.dart';
import 'package:stacked_chart/src/data/data.dart';

///Default color that is used when no color is passed
const barColors = [
  Color.fromARGB(255, 251, 138, 140),
  Color.fromARGB(255, 138, 143, 251),
  Color.fromARGB(200, 191, 214, 200),
];

class StackedChart extends StatelessWidget {
  /// Size of the widget itself. Please consider the
  /// value being passed as it need to hold given [barWidth]
  final Size size;

  /// Default value is 15
  final double barWidth;

  /// Buffer represent the space between the max value and
  /// max scale to represent the bar. Default value is 5
  final double buffer;
  final List<ChartData> data;
  final TextStyle? labelStyle;
  final MainAxisAlignment mainAxisAlignment;

  /// Whether to enable shadow behind the barchart to alter
  /// material look or flat glass like look
  final bool enableShadow;

  final bool showLabel;
  const StackedChart(
      {Key? key,
      required this.size,
      required this.data,
      this.labelStyle,
      this.buffer = 5,
      this.barWidth = 15,
      this.showLabel = false,
      this.enableShadow = true,
      this.mainAxisAlignment = MainAxisAlignment.spaceEvenly})
      : super(key: key);

  num get maxValue => data.getMaxValue;

  @override
  Widget build(BuildContext context) {
    return SizedBox.fromSize(
      size: size,
      child: Row(
        mainAxisAlignment: mainAxisAlignment,
        children: data
            .map(
              (value) => SizedBox.fromSize(
                size: Size(size.width / data.length, size.height),
                child: Bar(
                  enableShadow: enableShadow,
                  maxValue: maxValue + buffer,
                  size: Size(barWidth, size.height),
                  showLabel: showLabel,
                  data: value,
                  label: value.barLabel,
                  labelStyle: labelStyle ??
                      Theme.of(context).textTheme.caption ??
                      TextStyle(),
                ),
              ),
            )
            .toList(),
      ),
    );
  }
}

class Bar extends StatelessWidget {
  const Bar({
    Key? key,
    required this.size,
    required this.data,
    required this.maxValue,
    required this.labelStyle,
    required this.showLabel,
    required this.enableShadow,
    this.label,
  }) : super(key: key);

  final ChartData data;
  final Size size;
  final bool showLabel;
  final String? label;
  final num maxValue;
  final TextStyle labelStyle;
  final bool enableShadow;

  @override
  Widget build(BuildContext context) {
    return Column(
      mainAxisAlignment: MainAxisAlignment.center,
      children: [
        Expanded(
          flex: 8,
          child: SizedBox.fromSize(
            size: size,
            child: _BarStack(
              enableShadow: enableShadow,
              data: data,
              size: Size(size.width, size.height * (showLabel ? .8 : 1)),
              maxValue: maxValue,
            ),
          ),
        ),
        if (showLabel)
          Flexible(
              flex: 2,
              child: Column(
                children: [
                  SizedBox(height: 5),
                  Text(
                    label ?? '',
                    style: labelStyle,
                    overflow: TextOverflow.ellipsis,
                  ),
                ],
              ))
      ],
    );
  }
}

class _BarStack extends StatefulWidget {
  const _BarStack({
    Key? key,
    required this.data,
    required this.maxValue,
    required this.size,
    required this.enableShadow,
  }) : super(key: key);

  final num maxValue;

  final Size size;
  final ChartData data;
  final bool enableShadow;

  @override
  __BarStackState createState() => __BarStackState();
}

class __BarStackState extends State<_BarStack>
    with SingleTickerProviderStateMixin {
  late final AnimationController _controller;
  late Animation<double> _scale;

  @override
  void initState() {
    super.initState();
    _controller = AnimationController(
        vsync: this, duration: const Duration(milliseconds: 200));
    _scale = Tween<double>(begin: 1, end: .95).animate(_controller);
  }

  void _onTapDown(_) => _controller.forward();

  void _onTapUp(_) => _controller.reverse();

  void _onTapCancelled() => _controller.reverse();

  @override
  void dispose() {
    super.dispose();
    _controller.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      behavior: HitTestBehavior.opaque,
      onTap: widget.data.onPressed,
      onTapDown: _onTapDown,
      onTapUp: _onTapUp,
      onTapCancel: _onTapCancelled,
      child: AnimatedBuilder(
        animation: _controller,
        builder: (context, child) => Transform.scale(
          scale: _scale.value,
          child: Stack(alignment: Alignment.bottomCenter, children: [
            Container(
              child: _BarWidget(
                barColor: widget.data.barBackGroundColor ?? barColors.last,
                barSize: widget.size,
              ),
              decoration: BoxDecoration(
                borderRadius: BorderRadius.circular(8),
                boxShadow: [
                  BoxShadow(
                    offset: Offset(1, 1),
                    blurRadius: 4,
                    spreadRadius: 1,
                    color: Color.fromRGBO(
                        135, 135, 135, widget.enableShadow ? 1 : 0),
                  )
                ],
              ),
            ),
            ...widget.data.labelWithValue.entries
                .map(
                  (e) => Tooltip(
                    message: '${e.key.label}: ${e.value}',
                    child: _BarWidget(
                      barColor: e.key.color ??
                          barColors[widget.data.getIndexForGivenKey(e.key) %
                              barColors.length],
                      barSize: Size(
                          widget.size.width,
                          (widget.size.height / widget.maxValue) *
                              widget.data.getHeightValueForGivenKey(e.key)),
                    ),
                  ),
                )
                .toList()
          ]),
        ),
      ),
    );
  }
}

class _BarWidget extends StatelessWidget {
  const _BarWidget({
    Key? key,
    required this.barColor,
    required this.barSize,
  }) : super(key: key);

  final Color barColor;
  final Size barSize;

  @override
  Widget build(BuildContext context) {
    return Container(
      height: barSize.height,
      decoration: BoxDecoration(
        borderRadius: BorderRadius.circular(8),
        color: barColor,
      ),
      width: barSize.width,
    );
  }
}


example/lib/widgets/weekly_chart_demo.dart
import 'dart:math';

import 'package:flutter/material.dart';
import 'package:stacked_chart/stacked_chart.dart';

enum WeekDay { monday, tuesday, wednesday, thursday, friday, saturday, sunday }

class WeeklyChartDemo extends StatefulWidget {
  const WeeklyChartDemo({Key? key}) : super(key: key);

  @override
  _WeeklyChartDemoState createState() => _WeeklyChartDemoState();
}

class _WeeklyChartDemoState extends State<WeeklyChartDemo> {
  final rng = Random();
  List<BookingStatus> weeklyStatus = [];

  @override
  void initState() {
    createRandomWeeklyStatus();
    super.initState();
  }

  void createRandomWeeklyStatus() {
    weeklyStatus.clear();
    final List<BookingStatus> weeklyData = [];
    WeekDay.values
        .map(
          (day) => weeklyData.add(BookingStatus(
            dateTime: DateTime.now().add(
              Duration(days: day.index),
            ),
            bookings: {
              'Unfilled booking': rng.nextInt(20),
              'Filled booking': rng.nextInt(20),
            },
          )),
        )
        .toList();
    setState(() {
      weeklyStatus = weeklyData;
    });
  }

  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      onTap: createRandomWeeklyStatus,
      child: StackedChart(
        data: weeklyStatus,
        size: const Size(300, 150),
        showLabel: true,
      ),
    );
  }
}

class BookingStatus extends ChartData<LabelData, int>
    implements Comparable<BookingStatus> {
  final DateTime dateTime;
  final Map<String, int> bookings;

  static Map<LabelData, int> convertBookingToMapOfLabelDataInt(
      Map<String, int> bookings) {
    final Map<LabelData, int> convertedData = {};
    bookings.entries
        .map((e) => convertedData.addAll({LabelData(e.key): e.value}))
        .toList();
    return convertedData;
  }

  int get totalBookingCount =>
      bookings.values.reduce((total, value) => total = total + value);

  BookingStatus({required this.dateTime, this.bookings = const {}})
      : super(
            labelWithValue: convertBookingToMapOfLabelDataInt(bookings),
            barLabel: dateTime.day.toString());

  @override
  int compareTo(BookingStatus other) => dateTime.compareTo(other.dateTime);
}


example/lib/main.dart
import 'dart:math';

import 'package:flutter/material.dart';
import 'package:stacked_chart/stacked_chart.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Stacked Chart'),
      ),
      body: Center(
        child: WeeklyChartDemo(),
      ),
    );
  }
}

enum WeekDay { monday, tuesday, wednesday, thursday, friday, saturday, sunday }

class WeeklyChartDemo extends StatefulWidget {
  const WeeklyChartDemo({Key? key}) : super(key: key);

  @override
  _WeeklyChartDemoState createState() => _WeeklyChartDemoState();
}

class _WeeklyChartDemoState extends State<WeeklyChartDemo> {
  final rng = Random();
  List<BookingStatus> weeklyStatus = [];

  @override
  void initState() {
    createRandomWeeklyStatus();
    super.initState();
  }

  void createRandomWeeklyStatus() {
    weeklyStatus.clear();
    final List<BookingStatus> weeklyData = [];
    WeekDay.values
        .map(
          (day) => weeklyData.add(BookingStatus(
              dateTime: DateTime.now().add(
                Duration(days: day.index),
              ),
              bookings: {
                'Unfilled booking': rng.nextInt(20),
                'Filled booking': rng.nextInt(20),
              },
              onPressed: createRandomWeeklyStatus)),
        )
        .toList();
    setState(() {
      weeklyStatus = weeklyData;
    });
  }

  @override
  Widget build(BuildContext context) {
    return StackedChart(
      data: weeklyStatus,
      mainAxisAlignment: MainAxisAlignment.spaceBetween,
      size: const Size(300, 150),
      showLabel: true,
      enableShadow: true,
    );
  }
}

class BookingStatus extends ChartData<LabelData, int>
    implements Comparable<BookingStatus> {
  final DateTime dateTime;
  final Map<String, int> bookings;
  final VoidCallback? onPressed;

  static Map<LabelData, int> convertBookingToMapOfLabelDataInt(
      Map<String, int> bookings) {
    final Map<LabelData, int> convertedData = {};
    bookings.entries
        .map((e) => convertedData.addAll({LabelData(e.key): e.value}))
        .toList();
    return convertedData;
  }

  int get totalBookingCount =>
      bookings.values.reduce((total, value) => total = total + value);

  BookingStatus(
      {required this.dateTime, this.bookings = const {}, this.onPressed})
      : super(
          labelWithValue: convertBookingToMapOfLabelDataInt(bookings),
          barLabel: dateTime.day.toString(),
          onPressed: onPressed,
        );

  @override
  int compareTo(BookingStatus other) => dateTime.compareTo(other.dateTime);
}
